Mathlib/AlgebraicGeometry/Restrict.lean

```diff
@@ -34,55 +34,123 @@ variable {C : Type u₁} [Category.{v} C]
-    (X ∣_ᵤ U).presheaf.map (eqToHom <| U.openEmbedding.functor_obj_injective e).op := rfl
+lemma eq_presheaf_map_eqToHom {V W : Opens U} (e : U.ι ''ᵁ V = U.ι ''ᵁ W) :
+    X.presheaf.map (eqToHom e).op =
+      U.toScheme.presheaf.map (eqToHom <| U.openEmbedding.functor_obj_injective e).op := rfl
 
 @[simp]
-lemma opensRange_ιOpens {X : Scheme.{u}} (U : Opens X) : (Scheme.ιOpens U).opensRange = U :=
-  Opens.ext Subtype.range_val
+lemma nonempty_iff : Nonempty U.toScheme ↔ (U : Set X).Nonempty := by
+  simp only [toScheme_carrier, SetLike.coe_sort_coe, nonempty_subtype]
+  rfl
 
 attribute [-simp] eqToHom_op in
 /-- The global sections of the restriction is isomorphic to the sections on the open set. -/
 @[simps!]
-def Scheme.resTop (X : Scheme.{u}) (U : Opens X) : Γ(X ∣_ᵤ U, ⊤) ≅ Γ(X, U) :=
-  X.presheaf.mapIso (eqToIso U.openEmbedding_obj_top.symm).op
+def topIso : Γ(U, ⊤) ≅ Γ(X, U) :=
+  X.presheaf.mapIso (eqToIso U.ι_image_top.symm).op
+
+/-- The stalks of an open subscheme is isomorphic to the stalk of the original scheme. -/
+def stalkIso {X : Scheme.{u}} (U : X.Opens) (x : U) :
+    U.toScheme.presheaf.stalk x ≅ X.presheaf.stalk x.1 :=
+  X.restrictStalkIso (Opens.openEmbedding _) _
+
+@[reassoc (attr := simp)]
+lemma germ_stalkIso_hom {X : Scheme.{u}} (U : X.Opens)
+    {V : TopologicalSpace.Opens U} (x : V) :
+      U.toScheme.presheaf.germ x ≫ (U.stalkIso x.1).hom =
+        X.presheaf.germ ⟨x.1.1, show x.1.1 ∈ U.ι ''ᵁ V from ⟨x.1, x.2, rfl⟩⟩ :=
+    PresheafedSpace.restrictStalkIso_hom_eq_germ _ _ _ _ _
+
+end Scheme.Opens
+
+/-- If `U` is a family of open sets that covers `X`, then `X.restrict U` forms an `X.open_cover`. -/
+@[simps! J obj map]
+def Scheme.openCoverOfSuprEqTop {s : Type*} (X : Scheme.{u}) (U : s → X.Opens)
+    (hU : ⨆ i, U i = ⊤) : X.OpenCover where
+  J := s
+  obj i := U i
+  map i := (U i).ι
+  f x :=
+    haveI : x ∈ ⨆ i, U i := hU.symm ▸ show x ∈ (⊤ : X.Opens) by trivial
+    (Opens.mem_iSup.mp this).choose
+  covers x := by
+    erw [Subtype.range_coe]
+    have : x ∈ ⨆ i, U i := hU.symm ▸ show x ∈ (⊤ : X.Opens) by trivial
+    exact (Opens.mem_iSup.mp this).choose_spec
 
 /-- The open sets of an open subscheme corresponds to the open sets containing in the subset. -/
 @[simps!]
-def opensRestrict {X : Scheme.{u}} (U : Opens X) :
-    Opens (X ∣_ᵤ U) ≃ { V : Opens X // V ≤ U } :=
-  (IsOpenImmersion.opensEquiv (Scheme.ιOpens U)).trans (Equiv.subtypeEquivProp (by simp))
+def opensRestrict :
+    Scheme.Opens U ≃ { V : X.Opens // V ≤ U } :=
+  (IsOpenImmersion.opensEquiv (U.ι)).trans (Equiv.subtypeEquivProp (by simp))
 
-instance ΓRestrictAlgebra {X : Scheme.{u}} {Y : TopCat.{u}} {f : Y ⟶ X} (hf : OpenEmbedding f) :
-    Algebra (Scheme.Γ.obj (op X)) (Scheme.Γ.obj (op <| X.restrict hf)) :=
-  (Scheme.Γ.map (X.ofRestrict hf).op).toAlgebra
+instance ΓRestrictAlgebra {X : Scheme.{u}} (U : X.Opens) :
+    Algebra (Scheme.Γ.obj (op X)) (Scheme.Γ.obj (op U)) :=
+  (Scheme.Γ.map U.ι.op).toAlgebra
```

Comment:
```suggestion
instance ΓRestrictAlgebra {X : Scheme.{u}} (U : X.Opens) :
    Algebra Γ(X, ⊤) Γ(U, ⊤) :=
  (Scheme.Γ.map U.ι.op).toAlgebra
```

Reply:
This is copy pasted from elsewhere. I'll make this change in a follow-up PR.